Issue 8244

denisristic opened the pull request:
Updated Magento\Framework\Serialize\SerializerInterface.php to be able to take optional options.
Added options JSON_NUMERIC_CHECK to Magento\Framework\Controller\Result\Json.php

Example Magento Json Response: add product to cart
`{"cart":{"summary_count":19,"subtotal":"<span class=\"price\">$38.00<\/span>","possible_onepage_checkout":true,"items":[{"product_type":"simple","options":[],"qty":19,"item_id":"1","configure_url":"http:\/\/127.0.0.1:32769\/checkout\/cart\/configure\/id\/1\/product_id\/1\/","is_visible_in_site_visibility":true,"product_id":"1","product_name":"Test product","product_sku":"Test-product","product_url":"http:\/\/127.0.0.1:32769\/test-product.html","product_has_url":true,"product_price":"\n\n    <span class=\"price-excluding-tax\" data-label=\"Excl. Tax\">\n            <span class=\"minicart-price\">\n            <span class=\"price\">$2.00<\/span>        <\/span>\n\n        <\/span>\n","product_price_value":2,"product_image":{"src":"http:\/\/127.0.0.1:32769\/pub\/static\/version1489682122\/frontend\/Magento\/blank\/en_US\/Magento_Catalog\/images\/product\/placeholder\/thumbnail.jpg","alt":"Test product","width":"78","height":"78"},"canApplyMsrp":false}],"extra_actions":"","isGuestCheckoutAllowed":true,"website_id":"1","subtotal_incl_tax":"<span class=\"price\">$38.00<\/span>","subtotal_excl_tax":"<span class=\"price\">$38.00<\/span>","data_id":1489682301},"messages":{"messages":[{"type":"success","text":"You added Test product to your shopping cart."}],"data_id":1489682301}}`

new response:
`{"cart":{"summary_count":1,"subtotal":"<span class=\"price\">$2.00<\/span>","possible_onepage_checkout":true,"items":[{"product_type":"simple","options":[],"qty":1,"item_id":2,"configure_url":"http:\/\/127.0.0.1:32769\/checkout\/cart\/configure\/id\/2\/product_id\/1\/","is_visible_in_site_visibility":true,"product_id":1,"product_name":"Test product","product_sku":"Test-product","product_url":"http:\/\/127.0.0.1:32769\/test-product.html","product_has_url":true,"product_price":"\n\n    <span class=\"price-excluding-tax\" data-label=\"Excl. Tax\">\n            <span class=\"minicart-price\">\n            <span class=\"price\">$2.00<\/span>        <\/span>\n\n        <\/span>\n","product_price_value":2,"product_image":{"src":"http:\/\/127.0.0.1:32769\/pub\/static\/version1489682122\/frontend\/Magento\/blank\/en_US\/Magento_Catalog\/images\/product\/placeholder\/thumbnail.jpg","alt":"Test product","width":78,"height":78},"canApplyMsrp":false}],"extra_actions":"","isGuestCheckoutAllowed":true,"website_id":1,"subtotal_incl_tax":"<span class=\"price\">$2.00<\/span>","subtotal_excl_tax":"<span class=\"price\">$2.00<\/span>","data_id":1489685205},"messages":{"messages":[{"type":"success","text":"You added Test product to your shopping cart."}],"data_id":1489685205}}`

As can be seen all numbers are striped from quotes

Commit:
Merge branch 'develop' of github.com:denisristic/magento2 into develop

* 'develop' of github.com:denisristic/magento2:
  Revert "Added flags to static deploy command to skip filetypes (js, css, less, images, fonts...) and filter by theme"
  Added flags to static deploy command to skip filetypes (js, css, less, images, fonts...) and filter by theme

Commit:
Merge branch 'develop' of github.com:denisristic/magento2 into develop

* 'develop' of github.com:denisristic/magento2:

Commit:
UPDATED json & serialize serializer
UPDATED Magento Framework Controller Result Json encode function

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=8918) <br/>All committers have signed the CLA.

ishakhsuvarov commented:
#8244

ishakhsuvarov commented:
Hi @denisristic ,
Could you please look into Travis build failures?
Currently, it looks like integration tests detected some objects which became empty due to the change.
Thanks!

orlangur commented:
Jenkins? O.o

Commit:
Replaced \Zend_Json::encode with Json Serializer in Magento\Framework\Json\Encoder.php & Magento\Framework\DataObject.php.
There is still a lot of Zend_Json::encode calls, specially in zendframework1

Fixed code style

denisristic commented:
@ishakhsuvarov I've made some changes now but I don't see why any of my changes would return empty data. I'll dig deeper tomorrow

Commit:
Fixed encode -> serialize

Commit:
changed encode options

Commit:
removed hex_* encode options
updated test where expected assert was “int” to int

Commit:
removed all Zend_Json::encode & Zend_Json::decode from app & magento framework

Commit:
replaced all occurances of => ‘([0-9]+)’ to => int
fixed typo in dataobject

Commit:
fixed merge conflicts

denisristic committed:
Merge branch 'develop' into issue-8244

Commit:
reverted changes from last commit when number starts with 0

orlangur commented on app/code/Magento/Theme/Block/Adminhtml/Wysiwyg/Files/Tree.php:
```diff
@@ -57,7 +57,8 @@ public function getTreeLoaderUrl()
     public function getTreeJson($data)
     {
-        return \Zend_Json::encode($data);
+        $serializer = new \Magento\Framework\Serialize\Serializer\Json;
```
This is not correct. All dependencies shall be passed via constructor instead of direct instantiation.

Please fix all places.

orlangur commented:
Please rewrite this branch in such a way that all massive changes are placed in single commit, there are no merge commits and the builds are green.

Commit:
Changed Tree Class to use dependency via constructor instead direct initiation

denisristic commented:
@orlangur Thx, I've missed that. I've fixed Tree Class. If this build passes I'll create another branch with clean commits and create another pull request if that is OK?

denisristic commented:
@orlangur Unit tests are failing without any error message. Do you maybe know why?

orlangur commented:
>  I've fixed Tree Class.

Cool 👍 But you need to refactor all places with `new \Magento\Framework\Serialize\Serializer\Json`.

> If this build passes I'll create another branch with clean commits and create another pull request if that is OK?

There is no need for that, you can simply force push into the same branch ;)

> Do you maybe know why?

Because of https://github.com/magento/magento2/issues/8876, in temporary branch you can temporarily remove mentioned `FactoryTest`.

ishakhsuvarov commented:
@denisristic I would suggest working on the first version of this PR in a different way. I believe, that current changeset is way too large and may be backwards incompatible, which is an overkill for this kind of issue.

My suggestion would be: 

1. Start by adding an optional argument to the json serializer `\Magento\Framework\Serialize\Serializer\Json::serialize` with clear contract for the available options.
2. Refactor class initially reported in the issue (`Magento\Framework\Controller\Result\Json`) to use Json Serializer, rather then `Zend_Json`
3. Add an optional behavior for the `Result\Json` to operate without converting anything to string, but do not make it default. This way you will retain backward compatibility for all the existing Magento code and fix the problem described in the issue.

Thanks.

ishakhsuvarov commented:
Hi @denisristic 
I am closing this PR for now due to inactivity.
Please reopen if you decide to continue work on it.
Thank you

joshspivey commented:
@denisristic Do we have a fix for this yet?
